oas_worker/evaluate_devset.py

```diff
@@ -0,0 +1,313 @@
+#!/usr/bin/env python
+""" evaluate_devset
+
+Evaluate OAS Devset:
+  * Keywords
+  * Transcripts (TBD)
+
+Author(s): flipsimon, datadonk23
+Date: 10.01.22
+"""
+
+import httpx
+import argparse
+import pandas as pd
+from datetime import datetime
+from string import punctuation
+
+from app.jobs.spacy_pipe import spacy_load
+from devset.evaluate_devset_utils import (
+    precision_recall_f1, mean_average_precision_k, log_results
+)
+
+
+OAS_URL = 'http://admin:password@localhost:8080/api/v1'
+nlp = spacy_load("de_core_news_lg")
+
+
+def get_true_labels(fpath: str):
+    """
+    Gets ground truth labels from devset spreadsheet and lowercases them. It
+    returns labels as a dict with track ID as key and keywords list as values.
+
+    :param fpath: Filepath to devset spreadsheet
+    :return: List of Dicts {cba ID: keywords list}
+    :return list: list(dict(cba ID: [keywords]))
+    """
+    ground_truth = []
+    df = pd.read_csv(fpath)
+    df["cba_id"] = df["URL"].apply(
+        lambda s: "https://cba.fro.at/?p=" + s.split("/")[-1]
+    )
+    df["keywords"] = df.OAS_tags.str.lower().str.split(",").apply(
+        lambda x: [s.strip() for s in x]).tolist()
+    for _, row in df.iterrows():
+        ground_truth.append({row["cba_id"]: row["keywords"]})
+
+    return ground_truth
+
+
+def devsetIDs_to_cbaIDs(fpath: str):
+    """
+    Transforms IDs from devset spreadsheet to CBA IDs in OAS posts
+    (identifier field, not post ID!).
+
+    :param fpath: Filepath to devset spreadsheet
+    :return: List of OAS identifiers
+    """
+    df = pd.read_csv(fpath)
+    cba_head = "https://cba.fro.at/"
+    oas_param = "?p="
+    ids = df.URL.apply(lambda s: cba_head + oas_param + s.split("/")[-1])
+
+    return ids.to_list()
+
+
+def get_post_ids(cba_ids: list):
+    """
+    Transforms CBA IDs to OAS post IDs.
+
+    :param cba_ids: List of CBA identifiers
+    :return: Dict of CBA IDs as keys and OAS post IDs as values
+    """
+    url = f"{OAS_URL}/job"
+    job_ids = {}
+    oas_ids = {}
+
+    for cba_id in cba_ids:
+        body = {
+            "typ": "cba2oas_id",
+            "subjects": [""],
+            "args": {"identifier": cba_id}
+        }
+        res = httpx.post(url, json=body)
+        job_ids[cba_id] = str(res.json())
+
+    for cba_id, job_id in job_ids.items():
+        completed = False
+        job_url = url + f"/{job_id}"
+        while not completed:
```

Comment:
when doing requests in a loop, always add a timeout between the requests, e.g. `time.sleep(1)`